### PR TITLE
Codes and tests for tricky anonymous class field

### LINKED ISSUE #73: Crash in FieldDeclarationVisitor for field declaration inside anonymous class : cf-689
SPECIMIN crashes when casting `FieldDeclaration` inside anonymous class to `ClassOrInterfaceDeclaration`

Command: `./gradlew run --args='--outputDirectory "../ISSUES/CF/cf-689/output/" --root "../ISSUES/CF/cf-689/input/Issue689/" --targetFile "java/util/AbstractList.java" --targetMethod "java.util.SubList#listIterator(SubList<E>, int)"'`

Exception: 
```
Exception in thread "main" java.lang.ClassCastException: class com.github.javaparser.ast.expr.ObjectCreationExpr cannot be cast to class com.github.javaparser.ast.body.ClassOrInterfaceDeclaration (com.github.javaparser.ast.expr.ObjectCreationExpr and com.github.javaparser.ast.body.ClassOrInterfaceDeclaration are in unnamed module of loader 'app')
        at org.checkerframework.specimin.FieldDeclarationsVisitor.visit(FieldDeclarationsVisitor.java:35)
        at org.checkerframework.specimin.FieldDeclarationsVisitor.visit(FieldDeclarationsVisitor.java:19)
        at com.github.javaparser.ast.body.FieldDeclaration.accept(FieldDeclaration.java:116)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.lambda$visit$117(VoidVisitorAdapter.java:402)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at com.github.javaparser.ast.NodeList.forEach(NodeList.java:290)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.lambda$visit$118(VoidVisitorAdapter.java:402)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.visit(VoidVisitorAdapter.java:402)
        at com.github.javaparser.ast.expr.ObjectCreationExpr.accept(ObjectCreationExpr.java:118)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.lambda$visit$146(VoidVisitorAdapter.java:474)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.visit(VoidVisitorAdapter.java:474)
        at com.github.javaparser.ast.stmt.ReturnStmt.accept(ReturnStmt.java:84)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.lambda$visit$19(VoidVisitorAdapter.java:109)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at com.github.javaparser.ast.NodeList.forEach(NodeList.java:290)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.visit(VoidVisitorAdapter.java:109)
        at com.github.javaparser.ast.stmt.BlockStmt.accept(BlockStmt.java:76)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.lambda$visit$105(VoidVisitorAdapter.java:370)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.visit(VoidVisitorAdapter.java:370)
        at com.github.javaparser.ast.body.MethodDeclaration.accept(MethodDeclaration.java:104)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.lambda$visit$31(VoidVisitorAdapter.java:154)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at com.github.javaparser.ast.NodeList.forEach(NodeList.java:290)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.visit(VoidVisitorAdapter.java:154)
        at com.github.javaparser.ast.body.ClassOrInterfaceDeclaration.accept(ClassOrInterfaceDeclaration.java:98)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.lambda$visit$43(VoidVisitorAdapter.java:175)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at com.github.javaparser.ast.NodeList.forEach(NodeList.java:290)
        at com.github.javaparser.ast.visitor.VoidVisitorAdapter.visit(VoidVisitorAdapter.java:175)
        at com.github.javaparser.ast.CompilationUnit.accept(CompilationUnit.java:133)
        at org.checkerframework.specimin.SpeciminRunner.performMinimization(SpeciminRunner.java:129)
        at org.checkerframework.specimin.SpeciminRunner.main(SpeciminRunner.java:64)
```

Specimin [input](https://drive.google.com/file/d/1Uw9a_wntcQykbC36NIyToNoRJOaVHANo/view?usp=drive_link)

### PR DESCRIPTION
**Pull request by LoiNguyenCS**
Professor,

This is the PR to fix #73. I have replace the field `insideAnObjectCreation` with a method with the same name. I have also changed the way we get the simple name of a type (previously we used the `asString` method, which is not very accurate when there is a special symbol going with the type, such as [] or <>). And I have added codes to remove every class declaration that is not used by the target methods.

### CODE REVIEW

**Review by kelloggm — CHANGES_REQUESTED** [review]
This PR appears to contain code that is unrelated to fixing #73. Can you separate that code (i.e., the changes to how type variables are handled) into its own PR with its own test case? Once that is done, I can review this PR properly.

**Comment by LoiNguyenCS** [comment]
Professor,
I have removed the unrelated components. Please take a look. Thank you.

File: src/test/resources/ObjectCreationInsideAnonymousClass/expected/com/example/Simple.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+import org.testing.Baz;
+import org.testing.Foo;
+
+public class Simple {
+
+    public Baz getBaz() {
+        return new Baz() {
+
+            private Foo foo;
```
Comment (kelloggm) [inline src/test/resources/ObjectCreationInsideAnonymousClass/expected/com/example/Simple.java:11]: The initializer of this field is part of the target method `getBaz()`, so I wouldn't expect it to be removed here.

**Comment by LoiNguyenCS** [comment]
Professor,

I have updated the codes. Please take a look. Thank you.

File: src/main/java/org/checkerframework/specimin/TargetMethodFinderVisitor.java [hunk 2]
```diff
@@ -195,6 +195,7 @@ public Visitable visit(ConstructorDeclaration method, Void p) {
       insideTargetMethod = true;
       targetMethods.add(method.resolve().getQualifiedSignature());
       unfoundMethods.remove(methodName);
+      usedClass.add(method.resolve().getPackageName() + "." + method.resolve().getClassName());
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/specimin/TargetMethodFinderVisitor.java:198]: The code both here and on line 198 calls `method.resolve()` twice. This call is relatively expensive (it invokes the symbol solver), so we should call it once and store the result in a local variable.

**Review by kelloggm — APPROVED** [review]
One minor comment, but I don't need to review this again once that is fixed.

**Pull request merged**